sklearn/utils/stats.py

```diff
@@ -1,18 +1,36 @@
 import numpy as np
 
 from .extmath import stable_cumsum
+from sklearn.utils.fixes import _take_along_axis
 
 
 def _weighted_percentile(array, sample_weight, percentile=50):
     """
     Compute the weighted ``percentile`` of ``array`` with ``sample_weight``.
+    If ``array`` is 2D, compute weighted ``percentile`` along axis=0.
     """
-    sorted_idx = np.argsort(array)
+    n_dim = array.ndim
+    if n_dim == 0:
+        return array[()]
+    if array.ndim == 1:
+        array = array.reshape((-1, 1))
+    if array.shape != sample_weight.shape:
+        sample_weight = sample_weight.reshape(array.shape)
```

Comment:
Usually we expect `array` to be 2-D and `sample_weight` to be 1-D. It is going to blow up here:

```python
In [3]: # X is a 2-D array of 100 samples by 3 features 
   ...: X = np.random.randn(100, 3) 
   ...: # with its corresponding weight per samples 
   ...: sample_weight = np.random.randn(100) 
   ...:  
   ...: _weighted_percentile(X, sample_weight)                                                                                                                                                    
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-0716f7c3f0e0> in <module>
      4 sample_weight = np.random.randn(100)
      5 
----> 6 _weighted_percentile(X, sample_weight)

~/Documents/packages/scikit-learn/sklearn/utils/stats.py in _weighted_percentile(array, sample_weight, percentile)
     16         array = array.reshape((-1, 1))
     17     if array.shape != sample_weight.shape:
---> 18         sample_weight = sample_weight.reshape(array.shape)
     19     sorted_idx = np.argsort(array, axis=0)
     20     sorted_weights = _take_along_axis(sample_weight, sorted_idx, axis=0)

ValueError: cannot reshape array of size 100 into shape (100,3)

In [4]: _weighted_percentile(X, np.repeat(sample_weight.reshape(-1, 1), 3))
Out[4]: array([ 0.44473983, -0.38490779,  0.43873199])
```

One solution is to repeat `sample_weight` if `_num_samples(array) == _num_samples(sample_weight)`

Reply:
Thanks I didn't know this. Should I add a test for this case?